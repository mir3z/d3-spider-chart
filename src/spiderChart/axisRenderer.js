import noop from "lodash.noop";

/**
 * Renders axis for each statement in chart's data.
 * Each axis has it's name at the end.
 * Clicking axis name makes it highlighted.
 * @param parent SVG parent group
 * @param innerRadius inner radius at which each axis starts
 * @param outerRadius outer radius at which each axis ends
 * @param donutSize outer donut's size
 * @param labelSpacing additional spacing between axis end and axis label
 * @returns {function}
 */
export default function axisRenderer(parent, { innerRadius, outerRadius, donutSize, labelSpacing }) {
    const axesGroup = parent
        .append("g")
        .attr("class", CLASS_NAMES.AXIS_PARENT_GROUP);

    const setAxisLine = (x, y) => (el) => el
        .classed(CLASS_NAMES.AXIS_LINE_HIGHLIGHTED, d => d.highlight)
        .attr("x1", (d, i) => x(i, innerRadius))
        .attr("y1", (d, i) => y(i, innerRadius))
        .attr("x2", (d, i) => x(i, d.highlight ? outerRadius + donutSize : outerRadius))
        .attr("y2", (d, i) => y(i, d.highlight ? outerRadius + donutSize : outerRadius));

    const setLabel = (x, y, sliceAngle) => (el) => el
        .classed(CLASS_NAMES.AXIS_LABEL_HIGHLIGHTED, d => d.highlight)
        .classed("axis-label--dim", d => d.dim)
        .attr("x", (d, i) => x(i, outerRadius + labelSpacing))
        .attr("y", (d, i) => y(i, outerRadius + labelSpacing))
        .attr("text-anchor", (d, i) => {
            const angle = i * sliceAngle;

            if (angle > 0 && angle < Math.PI) {
                return "start";
            } else if (angle > Math.PI && angle < Math.PI * 2) {
                return "end";
            }

            return "middle";
        });

    const handleLabelClick = (data, update) => ({ axis }) => {
        const hasUnselected = data
            .map(series => series.values.find(s => s.axis === axis))
            .some(values => !values.focus);

        const anyHighlighted = data
            .map(series => series.values.find(s => s.axis === axis))
            .some(values => !values.highlight);

        const nextData = data
            .map(series => ({
                ...series,
                focus: false,
                values: series.values.map(v => v.axis === axis
                    ? { ...v, focus: hasUnselected, highlight: !v.highlight, dim: false }
                    : { ...v, focus: false, highlight: false, dim: anyHighlighted })
            }));

        update(nextData);
    };

    return function render(data = [], update = noop) {
        const axisList = data[0].values;
        const sliceAngle = Math.PI * 2 / axisList.length;
        const x = (i, r) => r * Math.sin(i * sliceAngle);
        const y = (i, r) => -r * Math.cos(i * sliceAngle);

        const axis = axesGroup
            .selectAll(`.${ CLASS_NAMES.AXIS }`)
            .data(axisList);

        axis.exit().remove();

        const enterAxis = axis.enter()
            .append("g")
            .attr("class", CLASS_NAMES.AXIS)
            .attr("data-axis", d => d.axis);

        enterAxis
            .append("line")
            .attr("class", CLASS_NAMES.AXIS_LINE)
            .call(setAxisLine(x, y));

        enterAxis
            .append("text")
            .attr("class", CLASS_NAMES.AXIS_LABEL)
            .text(d => d.axis)
            .call(setLabel(x, y, sliceAngle))
            .on("click", handleLabelClick(data, update));

        axis.select(`.${ CLASS_NAMES.AXIS_LINE }`)
            .call(setAxisLine(x, y));

        axis.select(`.${ CLASS_NAMES.AXIS_LABEL }`)
            .call(setLabel(x, y, sliceAngle))
            .on("click", handleLabelClick(data, update));

        return axesGroup;
    };
}

const CLASS_NAMES = {
    AXIS_PARENT_GROUP: "axis-group",
    AXIS: "axis",
    AXIS_LINE: "axis-line",
    AXIS_LINE_HIGHLIGHTED: "axis-line--highlighted",
    AXIS_LABEL: "axis-label",
    AXIS_LABEL_HIGHLIGHTED: "axis-label--highlighted"
};
