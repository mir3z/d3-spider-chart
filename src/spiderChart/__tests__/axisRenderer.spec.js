import axisRenderer from "../axisRenderer";
import { cleanup, findElementByText, setup } from "./utils";

describe("axis renderer", () => {

    afterEach(cleanup);

    const defaultOptions = { innerRadius: 0, outerRadius: 100, donutSize: 10, labelSpacing: 10 };
    const data = [{
        values: [
            { axis: "Statement 1" },
            { axis: "Statement 2" },
            { axis: "Statement 3", focus: true, highlight: true },
            { axis: "Statement 4" }
        ]
    }];

    it("renders its own group", () => {
        const svg = setup();
        const renderAxis = axisRenderer(svg, defaultOptions);

        const group = renderAxis(data);
        const ownGroup = svg.node().querySelector("g.axis-group");

        expect(ownGroup).not.toBeNull();
        expect(group.node()).toEqual(ownGroup);
    });

    it("renders axis for each statement", () => {
        const svg = setup();
        const renderAxis = axisRenderer(svg, defaultOptions);

        const group = renderAxis(data);

        expect(group.node()).toMatchSnapshot();
    });

    describe("when clicking axis label", () => {
        it("updates axis state if axis is NOT focused", () => {
            verifyAxisClick("Statement 2", [{
                focus: false,
                values: [
                    { axis: "Statement 1", focus: false, highlight: false, dim: true },
                    { axis: "Statement 2", focus: true, highlight: true, dim: false },
                    { axis: "Statement 3", focus: false, highlight: false, dim: true },
                    { axis: "Statement 4", focus: false, highlight: false, dim: true }
                ]
            }]);
        });

        it("updates axis state if axis is focused", () => {
            verifyAxisClick("Statement 3", [{
                focus: false,
                values: [
                    { axis: "Statement 1", focus: false, highlight: false, dim: false },
                    { axis: "Statement 2", focus: false, highlight: false, dim: false },
                    { axis: "Statement 3", focus: false, highlight: false, dim: false },
                    { axis: "Statement 4", focus: false, highlight: false, dim: false }
                ]
            }]);
        });

        it("re-renders axis based on new state", () => {
            const svg = setup();
            const renderAxis = axisRenderer(svg, defaultOptions);
            const update = updatedData => {
                const updatedGroup = renderAxis(updatedData);
                expect(updatedGroup.node()).toMatchSnapshot();
            };

            const group = renderAxis(data, update);

            findElementByText("Statement 4", group.node())
                .dispatchEvent(new Event("click"));
        });

        const verifyAxisClick = (axisLabel, expectedData) => {
            const svg = setup();
            const update = jest.fn();
            const renderAxis = axisRenderer(svg, defaultOptions);

            const group = renderAxis(data, update);

            findElementByText(axisLabel, group.node())
                .dispatchEvent(new Event("click"));

            expect(update).toHaveBeenCalledWith(expectedData);
        }
    });
});
