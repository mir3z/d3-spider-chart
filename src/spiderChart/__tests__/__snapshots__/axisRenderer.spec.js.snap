// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`axis renderer renders axis for each statement 1`] = `
<g
  class="axis-group"
>
  <g
    class="axis"
    data-axis="Statement 1"
  >
    <line
      class="axis-line"
      x1="0"
      x2="0"
      y1="0"
      y2="-100"
    />
    <text
      class="axis-label"
      text-anchor="middle"
      x="0"
      y="-110"
    >
      Statement 1
    </text>
  </g>
  <g
    class="axis"
    data-axis="Statement 2"
  >
    <line
      class="axis-line"
      x1="0"
      x2="100"
      y1="0"
      y2="-6.123233995736766e-15"
    />
    <text
      class="axis-label"
      text-anchor="start"
      x="110"
      y="-6.735557395310443e-15"
    >
      Statement 2
    </text>
  </g>
  <g
    class="axis"
    data-axis="Statement 3"
  >
    <line
      class="axis-line axis-line--highlighted"
      x1="0"
      x2="1.3471114790620885e-14"
      y1="0"
      y2="110"
    />
    <text
      class="axis-label axis-label--highlighted"
      text-anchor="middle"
      x="1.3471114790620885e-14"
      y="110"
    >
      Statement 3
    </text>
  </g>
  <g
    class="axis"
    data-axis="Statement 4"
  >
    <line
      class="axis-line"
      x1="0"
      x2="-100"
      y1="0"
      y2="1.8369701987210297e-14"
    />
    <text
      class="axis-label"
      text-anchor="end"
      x="-110"
      y="2.0206672185931328e-14"
    >
      Statement 4
    </text>
  </g>
</g>
`;

exports[`axis renderer when clicking axis label re-renders axis based on new state 1`] = `
<g
  class="axis-group"
>
  <g
    class="axis"
    data-axis="Statement 1"
  >
    <line
      class="axis-line"
      x1="0"
      x2="0"
      y1="0"
      y2="-100"
    />
    <text
      class="axis-label axis-label--dim"
      text-anchor="middle"
      x="0"
      y="-110"
    >
      Statement 1
    </text>
  </g>
  <g
    class="axis"
    data-axis="Statement 2"
  >
    <line
      class="axis-line"
      x1="0"
      x2="100"
      y1="0"
      y2="-6.123233995736766e-15"
    />
    <text
      class="axis-label axis-label--dim"
      text-anchor="start"
      x="110"
      y="-6.735557395310443e-15"
    >
      Statement 2
    </text>
  </g>
  <g
    class="axis"
    data-axis="Statement 3"
  >
    <line
      class="axis-line"
      x1="0"
      x2="1.2246467991473532e-14"
      y1="0"
      y2="100"
    />
    <text
      class="axis-label axis-label--dim"
      text-anchor="middle"
      x="1.3471114790620885e-14"
      y="110"
    >
      Statement 3
    </text>
  </g>
  <g
    class="axis"
    data-axis="Statement 4"
  >
    <line
      class="axis-line axis-line--highlighted"
      x1="0"
      x2="-110"
      y1="0"
      y2="2.0206672185931328e-14"
    />
    <text
      class="axis-label axis-label--highlighted"
      text-anchor="end"
      x="-110"
      y="2.0206672185931328e-14"
    >
      Statement 4
    </text>
  </g>
</g>
`;
